src/server/pfs/server/driver.go

```diff
@@ -4805,18 +4851,30 @@ func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_Put
 				pw.Close() // can't error
 			}
 			if req.Delete {
-				d.putFileLimiter.Acquire()
-				eg.Go(func() error {
-					defer d.putFileLimiter.Release()
+				if err := pl.start(req.File.Path); err != nil {
+					return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+				}
+				eg.Go(func() (retErr error) {
+					defer func() {
+						if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+							retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
+						}
+					}()
 					return f(req, nil)
 				})
 				continue
 			}
 			pr, pw = io.Pipe()
 			pr := pr
-			d.putFileLimiter.Acquire()
-			eg.Go(func() error {
-				defer d.putFileLimiter.Release()
+			if err := pl.start(req.File.Path); err != nil {
+				return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+			}
+			eg.Go(func() (retErr error) {
+				defer func() {
+					if err := pl.finish(req.File.Path); err != nil && retErr == nil {
```

Comment:
Is it possible for this call to `finish` and the one above to both be executed for a path? Is that okay?

Reply:
The goal (though I may have made a mistake) is that each `finish` call is tied to a `start` call above it, so as long as the sequence is `start(/a)->finish(/a)->start(/a)->finish(/a)`[1]  I think it should all be fine. If `finish` is called with no corresponding start then I think it would return an error when it couldn't find the path in its list of pending operations (which I think would cause `PutFile` to return an error)

[1] or, equivalently `start(/a)->start(/a) <block> -> finish(/a) -> start(/a) <unblock> -> finish(/a)`